lib/active_fulfillment/services/shipwire.rb

```diff
@@ -191,79 +191,88 @@ def parse_response(action, data)
+      end
+      items
     end
 
     def parse_inventory_response(xml)
-      response = {}
-      response[:stock_levels] = {}
-
-      document = REXML::Document.new(xml)
-      document.root.elements.each do |node|
-        if node.name == 'Product'
-          to_check = ['quantity']
-          to_check << 'pending' if include_pending_stock?
-
-          amount = to_check.sum { |a| node.attributes[a].to_i }
-          response[:stock_levels][node.attributes['code']] = amount
-        else
-          response[node.name.underscore.to_sym] = text_content(node)
-        end
+      response = { stock_levels: {} }
+      Parsing.with_xml_document(xml, response) do |document|
+        status = document.at_xpath('//Status'.freeze).child.content
+        total_products = document.at_xpath('//TotalProducts'.freeze).child.content
+        success = test? ? status == 'Test'.freeze : status == '0'.freeze
+        message = success ? 'Successfully received the stock levels'.freeze : document.at_xpath('//ErrorMessage'.freeze).child.content
+
+        {
+          status: status,
+          total_products: total_products,
+          stock_levels: compute_stock_levels(document),
+          message: message,
+          success: success
+        }
       end
+    end
 
-      response[:success] = test? ? response[:status] == 'Test' : response[:status] == '0'
-      response[:message] = response[:success] ? "Successfully received the stock levels" : message_from(response[:error_message])
-
-      response
+    def shipped_order?(node)
+      node.name == 'Order'.freeze && node.attributes['shipped'.freeze].text == 'YES'.freeze
     end
 
     def parse_tracking_response(xml)
-      response = {}
-      response[:tracking_numbers] = {}
-      response[:tracking_companies] = {}
-      response[:tracking_urls] = {}
-
-      document = REXML::Document.new(xml)
-      document.root.elements.each do |node|
-        if node.name == 'Order'
-          if node.attributes["shipped"] == "YES" && node.elements['TrackingNumber']
-            tracking_number = node.elements['TrackingNumber'].text.strip
-            response[:tracking_numbers][node.attributes['id']] = [tracking_number]
-
-            tracking_company = node.elements['TrackingNumber'].attributes['carrier']
-            response[:tracking_companies][node.attributes['id']] = [tracking_company.strip] if tracking_company
-
-            tracking_url = node.elements['TrackingNumber'].attributes['href']
-            response[:tracking_urls][node.attributes['id']] = [tracking_url.strip] if tracking_url
+      response = {
+        tracking_numbers: {},
+        tracking_companies: {},
+        tracking_urls: {}
+      }
+
+      Parsing.with_xml_document(xml, response) do |document, response|
+        document.root.try do |root_document|
+          root_document.elements.each do |node|
+            response[node.name.underscore.to_sym] = node.text.strip and next unless shipped_order?(node)
```

Comment:
You should pretty much always avoid `and` as it has extremely low prescience. I'd also like to see this line split out as the assignment, the boolean, the next, and then another boolean is pretty confusing.